# Chap 4.4 Bond Pricing

## Bond Pricing
* Suppose that a 2-Year Treasury bond with a principal of $100 provides coupons at the rate of 6% per annum semiannually.
* We calculate these bond price.

In [23]:
import math

In [43]:
pricipal = 100
bond_price = 0
mat_tenors = [0.5, 1.0, 1.5, 2.0]
zero_rates = [5.0, 5.8, 6.4, 6.8]  # zero rate(%)
coupon_rate = 0.06  
expiration = 2 # 2-Year Treasury Bond
per = 0.5 # Semi-Annually

In [44]:
for i, mat_tenor in enumerate(mat_tenors):
    if mat_tenor == expiration:
        bond_price += pricipal * (1 + coupon_rate * per) * math.exp( -1 * (zero_rates[i] * 0.01) * mat_tenor )
    else:
        bond_price += pricipal * (coupon_rate * per) * math.exp( -1 * (zero_rates[i] * 0.01) * mat_tenor )


In [45]:
print(bond_price)

98.3850627729396


## Bond Yield(채권수익률)
* A Bond's yield is the single discount rate that, we applied to all cash flows, give a bond price equals to its market price.
* Find $y$ satisfies the following equation.
$$
    3 \exp(-y * 0.5) + 3 \exp(-y * 1.0) + 3 \exp(-y * 1.5) + 103 \exp(-y * 2.0) = 98.3850627729396
$$
* We find $y$ using trial and error method.
 

In [51]:
max_trial = 10000
err_term = 0.000001
init_rate = 0.05
term = 0.0001

In [ ]:
def bond_formula(rate):
    return 3 * math.exp(-1 * rate * 0.5) + 3 * math.exp(-1 * rate * 1.0) + 3 * math.exp(-1 * rate * 1.5)+ 103 * math.exp(-1 * rate * 2.0)

In [55]:
for i in range(max_trial):
    result = bond_formula(init_rate)
    if math.fabs(result - bond_price) <= err_term:
        break
    else:
        if result > bond_price:
            init_rate += term
        else:
            init_rate -= term

print(f'Given bond, Bond Yield is {round(init_rate * 100, 5)} %.')


Given bond, Bond Yield is 6.76 %.


## Par Yield(액면가 수익률)
* The par yield for a certain bond maturity is the coupon rate that causes the bond price to equal its par value.
* Find $c$ satisfies the following equations.
$$
\frac{c}{2} \exp(-0.05 * 0.5) + \frac{c}{2} \exp(-0.058 * 1.0) + \frac{c}{2} \exp(-0.064 * 1.5) +( 100 + \frac{c}{2}) \exp(-0.068 * 2.0) = 100
$$
* We find $c$ using a straightforward way.

In [60]:
par_yield = 2 * (100 - 100 * math.exp(-0.068 * 2)) / (math.exp(-0.05 * 0.5) + math.exp(-0.058 * 1.0) + math.exp(-0.064 * 1.5) + math.exp(-0.068 * 2.0))
print(f'Par yield is {round(par_yield, 4)} %')

Par yield is 6.8729 %
